### PR TITLE
Chore – Integration tests for atlas web core

### PR DESCRIPTION
**Pull request by alfonsomunozpomer**
Manual checks

- [x] This branch is passing the CI tests as part of a pull request in the atlas-web-bulk or atlas-web-scxa repos and I have indicated that PR here. 

This PR includes a .env file to run integration tests, and some fixes to a few tests that were failing after some time of neglect. See also https://github.com/ebi-gene-expression-group/atlas-web-single-cell/pull/346 and https://github.com/ebi-gene-expression-group/atlas-web-bulk/pull/154 for context.

### CODE REVIEW

File: src/main/java/uk/ac/ebi/atlas/utils/EuropePmcClient.java [hunk 1]
```diff
@@ -33,6 +35,13 @@ public Optional<Publication> getPublicationByDoi(String doi) {
     }
 
     public Optional<Publication> getPublicationByPubmedId(String pubmedId) {
+        // Currently a query for an empty PubMed ID incorrectly returns a result; you can try it yourself:
+        // https://www.ebi.ac.uk/europepmc/webservices/rest/search?query=SRC:MED%20AND%20EXT_ID:&format=json
+        // "resultList":{"result":[{"id":"35811804","source":"MED","pmid":"35811804","pmcid":"PMC9218589"...
+        // You can remove this check and this comment if EuropePmc fix this bug
```
Comment (ke4) [inline src/main/java/uk/ac/ebi/atlas/utils/EuropePmcClient.java:42]: @alfonsomunozpomer Has this bug been reported to EuropePMC, yet?
Would be better to add a bug ID to this comment. That way a person encounter this class could check if that has been fixed yet, or not.

File: src/test/java/uk/ac/ebi/atlas/bioentity/properties/BioEntityCardPropertiesIT.java [hunk 2]
```diff
@@ -77,26 +88,26 @@ public void wbpsAndEnsemblPropertiesAreEquivalent() {
...
+    void ensemblFamilyTemplateDoesNotUseThePropertyValueButTheGeneId() {
         Species species = speciesFactory.create("mus musculus");
         assertThat(BioEntityCardProperties.getUrlTemplate(ENSFAMILY_DESCRIPTION, species))
                 .doesNotContain("{0}")
                 .contains("{1}");
     }
 
-    @Test
-    public void bioentityPropertyWithNoUrl() {
+    @RepeatedTest(100)
```
Comment (ke4) [inline src/test/java/uk/ac/ebi/atlas/bioentity/properties/BioEntityCardPropertiesIT.java]: Is this intentionally added or accidentally left here? Thanks

File: src/test/java/uk/ac/ebi/atlas/solr/analytics/query/AnalyticsQueryClientIT.java [hunk 3]
```diff
@@ -213,30 +74,154 @@ void returnsAllResultsFound() {
...
         }
 
-        @Override
-        protected String fetchResponseAsString(String url, SolrQuery q) {
-            return URLDecoder.decode(q.toString(), StandardCharsets.UTF_8);
+        @Test
+        void queryWithCategory() {
+            var queryMade = subject.queryBuilder()
+                    .bioentityIdentifierFacets(-1)
+                    .queryIdentifierSearch(SemanticQuery.create(SemanticQueryTerm.create("GO:1234567", "go")))
```
Comment (ke4) [inline src/test/java/uk/ac/ebi/atlas/solr/analytics/query/AnalyticsQueryClientIT.java]: Is it possible to replace all of the magic words with randomly generated values?
I meant these ones: `GO:1234567`, `go`, `E-MTAB-513`.
Thanks

File: src/test/java/uk/ac/ebi/atlas/solr/analytics/query/AnalyticsQueryClientIT.java [hunk 4]
```diff
@@ -213,30 +74,154 @@ void returnsAllResultsFound() {
...
         }
 
-        @Override
-        protected boolean responseNonEmpty(String jsonFromSolr) {
-            return true;
+        @Test
+        void queryWithNoCategoryButObviouslyAnEnsemblIdDoesABioentityIdentifierQuery() {
+            var queryMade = subject.queryBuilder()
+                    .bioentityIdentifierFacets(-1)
+                    .queryIdentifierSearch(SemanticQuery.create(SemanticQueryTerm.create("ENSG00000006432")))
```
Comment (ke4) [inline src/test/java/uk/ac/ebi/atlas/solr/analytics/query/AnalyticsQueryClientIT.java]: magic words ^^^

File: src/test/java/uk/ac/ebi/atlas/solr/analytics/query/AnalyticsQueryClientIT.java [hunk 5]
```diff
@@ -213,30 +74,154 @@ void returnsAllResultsFound() {
...
+            //assertThat(queryMade).doesNotContain("identifier_search");
+            assertThat(queryMade).contains("bioentity_identifier");
+
+        }
+
+        @Test
+        void speciesComeInQuoted() {
+            var queryMade = subject.queryBuilder()
+                    .bioentityIdentifierFacets(-1)
+                    .ofSpecies("oryza sativa")
```
Comment (ke4) [inline src/test/java/uk/ac/ebi/atlas/solr/analytics/query/AnalyticsQueryClientIT.java]: magic words ^^^

File: src/test/java/uk/ac/ebi/atlas/solr/analytics/query/AnalyticsQueryClientIT.java [hunk 6]
```diff
@@ -213,30 +74,154 @@ void returnsAllResultsFound() {
...
+                    .fetch();
+
+            assertThat(queryMade).contains("\"oryza sativa\"");
+        }
+
+        @Test
+        void weGuessThatZincFingerCanNotBeAKeyword() {
+            var queryMade = subject.queryBuilder()
+                    .bioentityIdentifierFacets(-1)
+                    .queryIdentifierSearch(SemanticQuery.create("zinc finger"))
```
Comment (ke4) [inline src/test/java/uk/ac/ebi/atlas/solr/analytics/query/AnalyticsQueryClientIT.java]: magic words ^^^

File: src/test/java/uk/ac/ebi/atlas/solr/analytics/query/AnalyticsQueryClientIT.java [hunk 7]
```diff
@@ -213,30 +74,154 @@ void returnsAllResultsFound() {
...
+        void defaultQueryIsTheSolrDefault() {
+            var queryMade = subject.queryBuilder().fetch();
+            assertThat(queryMade).contains("q=*:*");
+        }
+
+        @Test
+        void omitEmptyConditionQuery() {
+            var queryMade = subject.queryBuilder()
+                    .bioentityIdentifierFacets(-1)
+                    .queryIdentifierSearch(SemanticQuery.create("zinc finger"))
```
Comment (ke4) [inline src/test/java/uk/ac/ebi/atlas/solr/analytics/query/AnalyticsQueryClientIT.java]: magic words ^^^

File: src/test/java/uk/ac/ebi/atlas/solr/analytics/query/AnalyticsQueryClientIT.java [hunk 8]
```diff
@@ -213,30 +74,154 @@ void returnsAllResultsFound() {
...
+            assertThat(queryMade).contains("E-MTAB-513");
+
+            assertThat(queryMade).doesNotContain("conditionsSearch");
+        }
+
+        @Test
+        void bothConditionQueryAndIdentifierSearchMakeItIntoTheQueryString() {
+            var queryMade = subject.queryBuilder()
+                    .bioentityIdentifierFacets(-1)
+                    .queryIdentifierSearch(SemanticQuery.create("zinc finger"))
```
Comment (ke4) [inline src/test/java/uk/ac/ebi/atlas/solr/analytics/query/AnalyticsQueryClientIT.java]: magic words ^^^

File: src/test/java/uk/ac/ebi/atlas/solr/analytics/query/AnalyticsQueryClientIT.java [hunk 9]
```diff
@@ -213,30 +74,154 @@ void returnsAllResultsFound() {
...
+
+            assertThat(queryMade).contains("conditions_search");
+            assertThat(queryMade).contains("liver");
+        }
+
+        @Test
+        void queryConditionSearchOrIdentifierSearchIncludesTheQueryStringTwiceForQueriesWithNoCategory() {
+            var queryMade = subject.queryBuilder()
+                    .bioentityIdentifierFacets(-1)
+                    .queryIdentifierOrConditionsSearch(SemanticQuery.create("tasty pancake"))
```
Comment (ke4) [inline src/test/java/uk/ac/ebi/atlas/solr/analytics/query/AnalyticsQueryClientIT.java]: magic words ^^^

File: src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java [hunk 10]
```diff
@@ -1,26 +1,24 @@
...
-public class EuropePmcClientIT {
+class EuropePmcClientIT {
     @Inject
     private EuropePmcClient subject;
 
     @Test
-    public void publicationForValidDoi() {
-        Optional<Publication> result = subject.getPublicationByDoi("10.1126/sciimmunol.aan8664");
+    void publicationForValidDoi() {
+        var result = subject.getPublicationByDoi("10.1126/sciimmunol.aan8664");
```
Comment (ke4) [inline src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java]: Sorry, but if it is possible to mock it, then we should. Removes the possibility to have a broken test if `EuropePMC` removes that publication.

File: src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java [hunk 11]
```diff
@@ -33,8 +31,8 @@ public void publicationForValidDoi() {
     }
 
     @Test
-    public void publicationForValidPubmedId() {
-        Optional<Publication> result = subject.getPublicationByPubmedId("29352091");
+    void publicationForValidPubmedId() {
+        var result = subject.getPublicationByPubmedId("29352091");
```
Comment (ke4) [inline src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java]: magic word ^^^

File: src/test/resources/fixtures/gxa/experiment.sql [hunk 12]
```diff
@@ -16,7 +16,7 @@ INSERT INTO experiment (accession, type, access_key, last_update, pubmed_ids, do
...
 INSERT INTO experiment (accession, type, access_key, last_update, pubmed_ids, dois, private, species, load_date) VALUES ('E-MTAB-5214', 'RNASEQ_MRNA_BASELINE', '57e2cba9-be39-4ba7-b72f-1e204653f925', '2019-12-14 01:36:08.171973', '25954001', '10.1126/science.1262110', FALSE, 'Homo sapiens', '2019-12-14 01:36:08.171973');
-INSERT INTO experiment (accession, type, access_key, last_update, pubmed_ids, dois, private, species, load_date) VALUES ('E-MTAB-5423', 'RNASEQ_MRNA_BASELINE', '0f54bd3e-633d-4431-be0e-f0f0c7e762b3', '2019-12-14 01:35:08.184186', NULL, NULL, FALSE, 'Homo sapiens', '2019-12-14 01:35:08.184186');
+-- INSERT INTO experiment (accession, type, access_key, last_update, pubmed_ids, dois, private, species, load_date) VALUES ('E-MTAB-5423', 'RNASEQ_MRNA_BASELINE', '0f54bd3e-633d-4431-be0e-f0f0c7e762b3', '2019-12-14 01:35:08.184186', NULL, NULL, FALSE, 'Homo sapiens', '2019-12-14 01:35:08.184186');
```
Comment (ke4) [inline src/test/resources/fixtures/gxa/experiment.sql]: We either need to remove this line or add it back without commented it out.

**Review by ke4 — CHANGES_REQUESTED** [review]
It looks good, but I added some comments related to magic word(s)/number(s).
I know it was most probably not added by you, but as you changed them you just won the option to fix this issue. Sorry about that. 😊

File: src/main/java/uk/ac/ebi/atlas/utils/EuropePmcClient.java (hunk 1 above)
Comment (alfonsomunozpomer) [reply src/main/java/uk/ac/ebi/atlas/utils/EuropePmcClient.java:42]: I reported via [EuropePMC’s helpdesk](https://europepmc.org/ContactUs) so, as far as I know, there’s no external ID we can check. The helpdesk ticket ID from the reply is EuropePMC #657887.

Comment (ke4) [reply src/main/java/uk/ac/ebi/atlas/utils/EuropePmcClient.java:42]: I created a message regarding this issue here: https://groups.google.com/a/ebi.ac.uk/g/epmc-webservices.
I guess we could resolve this conversation now and whenever we got a response to my message we could just create a simple PR with a commit with the addition of the ticket URL

File: src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java [hunk 13]
```diff
@@ -1,54 +1,117 @@
...
 
-        assertThat(result.orElseThrow(RuntimeException::new))
-                .extracting("doi", "authors", "title")
+
+        var result = subject.getPublicationByDoi(doi);
+
+        assertThat(result)
+                .get()
+                // Value set in JSON response
+                .hasFieldOrPropertyWithValue("doi", "10.1126/sciimmunol.aan8664")
```
Comment (ke4) [inline src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java]: This bit is a little weird to me as we are trying to get a publication by a generated random DOI, then we check if the response contains not that random DOI, but a hardcoded one.

File: src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java [hunk 14]
```diff
@@ -1,54 +1,117 @@
...
+                .andRespond(withSuccess(pubmedIdResponse, MediaType.APPLICATION_JSON));
 
-        assertThat(result.orElseThrow(RuntimeException::new))
-                .extracting("pubmedId", "authors", "title")
+        var result = subject.getPublicationByPubmedId(pubmedId);
+
+        assertThat(result)
+                .get()
+                // Value set in JSON response
+                .hasFieldOrPropertyWithValue("pubmedId", "29352091")
```
Comment (ke4) [inline src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java]: The same as above ^^^

File: src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java [hunk 15]
```diff
@@ -1,54 +1,117 @@
...
+
+    @Test
+    void noResultForBlankPubmedId() {
+        // Because of a bug in EuropePMC, our client doesn’t make a request to EuropePMC if the pubmed ID is blank.
+        // Uncomment the following block and mockServer.verify() at the end if the bug is fixed:
+        // mockServer.expect(once(), requestTo(startsWith(EuropePmcClient.URL)))
+        //         .andExpect(method(HttpMethod.GET))
+        //         .andExpect(queryParam("format", URLEncoder.encode("json")))
+        //         .andExpect(queryParam("query", URLEncoder.encode("SRC:MED AND EXT_ID:")))
+        //         .andRespond(withSuccess(EMPTY_RESPONSE, MediaType.APPLICATION_JSON));
```
Comment (ke4) [inline src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java]: Is the `EMPTY_RESPONSE` constant value define somewhere? I can't see it. Maybe my bad.

File: src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java [hunk 16]
```diff
@@ -1,54 +1,117 @@
...
+                .andRespond(withSuccess(noResultsResponse, MediaType.APPLICATION_JSON));
 
-        assertThat(result1.isPresent()).isFalse();
-        assertThat(result2.isPresent()).isFalse();
+        assertThat(subject.getPublicationByDoi("")).isEmpty();
+
+        mockServer.verify();
     }
+
+    @Test
```
Comment (ke4) [inline src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java:96]: Shall we disable/ignore this test as it is not adding any value for now?

**Review by ke4 — COMMENTED** [review]
I added some comments. We can discuss it on today's standup.
Thanks for your work on this.

File: src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java (hunk 15 above)
Comment (alfonsomunozpomer) [reply src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java]: This is from an earlier version of the code. I’ll replace it with the `europepmc-no-results-response.json` resource.

File: src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java (hunk 16 above)
Comment (alfonsomunozpomer) [reply src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java:96]: It still checks a code path that exists in the main classes, so I think we should keep it.

File: src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java (hunk 13 above)
Comment (alfonsomunozpomer) [reply src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java]: As agreed, I changed the test to disregard the random DOI/PubMed ID and just check that the retrieved fields from the example JSON response aren’t empty.

File: src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java (hunk 14 above)
Comment (alfonsomunozpomer) [reply src/test/java/uk/ac/ebi/atlas/utils/EuropePmcClientIT.java]: The same as above ^^^ :P

**Review by ke4 — APPROVED** [review]
Look perfect. 💯 👍 Thanks for all your hard work.

**Pull request merged**